### PR TITLE
Cron dailies fix (dailies are reset with no damage when user is at the Inn)

### LINKED ISSUE #5070: Make Dailies Refresh While in Inn
When the Inn was originally devised, Dailies were frozen because we wanted to discourage cheating. Since then, we've taken a much more laissez-faire approach to cheating (it's only cheating if it's not motivating to you, and even then, you're only cheating yourself), and we've also gotten a lot of feedback from users who find it frustrating that their Dailies don't reset while they are resting in the inn. This sometimes leads to people leaving HabitRPG during a time when it could be most helpful, for example during finals or a flare-up of illness, or forces them to use cumbersome work-arounds when they are already stressed or sick.

The only change to make here would be to make the Dailies refresh every day, without causing damage, while you are resting in the Inn. You'd still get full credit for checking them off (as if you were checking a grey Daily) and you still wouldn't attack a boss or collect quest items until you checked out of the inn.

### PR DESCRIPTION
**Pull request by hairlessbear**
Fixes #5070.

### CODE REVIEW

**Comment by hairlessbear** [comment]
Huh...looks like the commits from my other PR were included here. I'm...honestly not sure why or how that happened. I'll try to fix it, although I'm not quite sure how to do that off the top of my head.

Edit: Sorry, don't have time to figure it out at the moment. Will try to remove the erroneous commits later.

**Comment by crookedneighbor** [comment]
No worries, it happens. Probably you checked out your new branch while on the other pr branch, rather than from your develop branch, so it created the new one off of your pr branch. <- very confusing sentence

Trying to figure out the easiest way for you to remove those commits...

``` bash
# Assuming you've added an upstream remote, this gets the latest changes from habitrpg
git fetch upstream develop
# This resets your branch to the version on habitrpg's develop branch
git reset --hard upstream/develop
# This takes the commit of your change to cron and adds it to the reset branch
git cherry-pick 6ffa1936345e1a9b1ebee3040fa60baf50694ce6
# Overwrites your remote branch with your new version of the branch
git push --force
```

If ^ that doesn't work for you, you can always check out a new branch off of develop, redo the change, and submit a new pull request.

**Comment by Alys** [comment]
I've copied this code to the beta site and I'll give it a try over the next few days to see how it feels to be able to use Dailies while Resting. If anyone else wants to try it, remember that for the code to take effect, your account's first action of the day MUST be on the beta website (that includes any actions from automated processes or third-party tools).

Note that the beta site also contains many changes to the skills and other things (mostly mana-related): https://github.com/HabitRPG/habitrpg/issues/3029#issuecomment-68128382

**Comment by Alys** [comment]
Note also that the beta site contains many changes to the skills and other things (mostly mana-related): https://github.com/HabitRPG/habitrpg/issues/3029#issuecomment-68128382

**Comment by hairlessbear** [comment]
@crookedneighbor 

> Probably you checked out your new branch while on the other pr branch, rather than from your develop branch, so it created the new one off of your pr branch.

Nailed it. I didn't realize you could create branches off of branches (although in retrospect, I can't think of any reason why you _wouldn't_ be able to do that. Oh well; learn something new every day.) 

And as you can see, your fix worked; thank you!

**Comment by crookedneighbor** [comment]
@hairlessbear There are two basic ways to checkout a branch on the command line:

``` bash
git checkout -b <branch_name>
```

Which creates a new branch based on the branch you are currently on, or:

``` bash
git checkout -b <branch_name> <branch_you_are_basing_off_of>
```

Which as the name suggests, creates a new branch off of another branch. So, if you want to make sure you're up to date with upstream develop each time you branch off, you could do this:

``` bash
git fetch upstream
git checkout -b <branch_name> upstream/develop
```

And that'll make a new branch off of upstream/develop. One caveat with that method is that it'll set the origin for that branch to the upstream repository, so you'll also need to run:

```
git push -u origin <branch_name>
```

the first time you push, so that it creates the same remote branch on your own fork of Habit and let you make a pull request. The -u flag tells it to remember that remote branch as the origin for this branch. Git is magical but complicated. :smile:

**Comment by crookedneighbor** [comment]
I should also say, I'd like some testing around this new behavior before it gets merged in. 

If you don't feel up to the task, no worries, we can either work on them together, or I can work on them alone.

**Comment by deilann** [comment]
Tasks did not reset.

First action of the day was on Beta.

**Comment by hairlessbear** [comment]
@crookedneighbor I'd like to write the tests, but I haven't written any before, so I have no idea how to do it or where to start. If you're willing to help me learn how to write them, I'd love to help, but if that's an unnecessary amount of work for you (which would be completely understandable!), then it would probably be best if you wrote them on your own.

@deilann Well that sucks; it worked perfectly when I tested it locally. Does anyone have any hypotheses why it would work locally but not on the beta?

**Comment by deilann** [comment]
Might have been that I hadn't loaded the new code yet. Window was stale.

**Comment by deilann** [comment]
I'm planning on remaining in the inn today as well, and we should get more
data soon to find out if I'm an outlier. (aka when Alys wakes up)

**Comment by hairlessbear** [comment]
I created a test account (creatively named hairlessbeartestuser) that I'm testing this with. It's currently in the inn with multiple dailies checked, and cron is set to run in 13 minutes. We'll see what happens.

**Comment by deilann** [comment]
_waits with baited breath_

No really. I'm eating fish balls. And really want this to actually work.

**Comment by hairlessbear** [comment]
Curses, it didn't work. But I'm not sure cron ran. I checked dailies history in the DDT, and there isn't any, indicating that cron didn't run (right?) Just to double-check, I added a `+` only habit and hit it once, so on cron run, it'll revert back to yellow. Cron is set for the nearest hour :)

**Comment by deilann** [comment]
Cron won't always run if you set it that soon, just so you know.

**Comment by deilann** [comment]
Also, cron definitely ran for me. My drop cap was reset.

**Comment by deilann** [comment]
Also, also, I didn't get any Dailies history, due to being in the inn.
Sounds like it's not working for some reason.

**Comment by hairlessbear** [comment]
Hmm. I'll set cron for a few hours from now, just to be safe. Thanks for the tip.

**Comment by Alys** [comment]
It is entirely possible that I did not push the change to beta correctly. I
will be looking into that a bit later.

**Comment by Alys** [comment]
Yes, it was definitely me not pushing to beta correctly. The cron change works for me there now.

**Comment by hairlessbear** [comment]
Just tested it with my test user, and it worked for me too.

**Comment by hairlessbear** [comment]
@crookedneighbor regarding testing, after some poking around, it seems to me that the best place for the test is in `test/common/algos.mocha.coffee` (maybe by modifying/using [expectedDayResetNoDamage](https://github.com/HabitRPG/habitrpg/blob/10a3c0094b6f6797458669e32e8225bb65fbb631/test/common/algos.mocha.coffee#L106-122)). But according to the test readme, the tests in that file aren't used. Is the readme accurate? I can't find any other tests in the named areas that replicate the functionality in `algos.mocha.coffee`.

**Comment by crookedneighbor** [comment]
That readme is not accurate. The tests in the file you pointed out are being used and you are exactly correct, that is the file that needs to be edited. Specifically, a new `expect` should be added to the [dailies section](https://github.com/HabitRPG/habitrpg/blob/10a3c0094b6f6797458669e32e8225bb65fbb631/test/common/algos.mocha.coffee#L108-L112).

**Comment by crookedneighbor** [comment]
Please revert the changes you've made to the language files, other than the basic English ones. Those files are managed by [Transifex](https://www.transifex.com/) and not through Github.

**Comment by hairlessbear** [comment]
Ok, sorry about that. When you say "basic English ones", I'm assuming you mean just `/en` and `/en_GB`, correct?

**Comment by Alys** [comment]
Just en. See the readme-first document in each of the other language directories. :)

**Comment by hairlessbear** [comment]
Heh, well that's embarrassing. I'll revert those right away.

**Comment by Alys** [comment]
That happens often. :)  If you can think of a way to make it more obvious, please tell us!

**Comment by hairlessbear** [comment]
That took way more commits than it should have (vim added EOF newlines that were annoying to remove), but it's all back to normal now. 

@crookedneighbor I'll try to write the tests; we'll see how it goes. Thanks for the clarification.

**Comment by crookedneighbor** [comment]
@hairlessbear ping me if you're having trouble. Coffeescript can be a little confusing.

**Comment by hairlessbear** [comment]
@crookedneighbor or anyone else: is there information somewhere on how to test locally? I've found bits and pieces crawling through the code and comments on various PRs, but I haven't found anything cohesive. (@Alys, we should definitely add testing information to the wiki).

I know that the tests are run via `npm test`, but when I try that, I get `Error: Timed out waiting for Selenium`. I tried running `TRAVIS=true npm test`, as described [here](https://github.com/HabitRPG/habitrpg/pull/4873#issuecomment-86331968), but that didn't fix it either.

**Comment by crookedneighbor** [comment]
@hairlessbear Hmmm... Do you have Java installed on your system? Selenium relies on Java. 

Running `npm test` runs the entire test suite (which will take a long time), you could just run the file you want to change by running `mocha test/common/algos.mocha.coffee`. You'll need to have mocha installed first (`npm install -g mocha`)

I don't think that section of the tests requires selenium. 

I'm looking over the common test file again, and I'm suddenly not convinced that the function you pointed out is the best place to modify. I don't have a lot of experience with this file though, so I could be wrong. I did notice that most of the tests pertaining to dailies have been set to skip... so some work may need to be done on this file before a good test _can even be written_.

**Comment by hairlessbear** [comment]
I'm running the VM through vagrant, and apparently it didn't come with java installed. After installing `default-jre-headless`, `npm test` ran, but with the end-to-end tests failing due to Firefox not being installed. I can deal with that another time.

I'm not sure that exact function is perfect, but I do think that the file is right (maybe the tests should be in [this section](https://github.com/HabitRPG/habitrpg/blob/10a3c0094b6f6797458669e32e8225bb65fbb631/test/common/algos.mocha.coffee#L140-L187)?). Nowhere else I looked had any tests anywhere close to what I'm testing. Also, as far as I can tell, there are no tests anywhere that test the Inn functionality. Seems like that'll have to be added in with the tests for this.

Another question: how can you tell which tests are set to skip? 

And I'm sorry for the barrage of simple questions; I'm _really_ new to this area. As soon as I get a handle on how testing works, I'm going to make sure to update the wiki so you guys will be able to point somewhere for answers instead of having to answer the same questions over and over.

**Comment by gisikw** [comment]
Yep, at the moment, e2e will fail in headless envs. But the summary should let you know whether the other suites are passing :)

Seeing skipped tests - they should show up as pending if you scroll up (and are probably blue)

**Comment by gisikw** [comment]
As far as the code to skip, either `xdescribe` or `it.skip`

**Comment by crookedneighbor** [comment]
@hairlessbear You're definitely right in the file choice, that's the place to write the test. I'm just not sure the specific function you first pointed out is the best place to add it. I don't see any tests for behavior related to resting in the inn, so it's probably best to write a new block of tests within the user block you just posted for how the user should behave when resting in the inn.

### RE: skipped tests.

When running `mocha test/common/algos.mocha.coffee`, you'll see this:

screen shot 2015-04-25 at 3 08 34 pm

The blue tests are the ones that are being skipped. If you look at [line 617](https://github.com/HabitRPG/habitrpg/blob/10a3c0094b6f6797458669e32e8225bb65fbb631/test/common/algos.mocha.coffee#L617), you'll see that the whole describe block there is being skipped.

**Comment by hairlessbear** [comment]
@giskw @crookedneighbor Thanks for the help! I'll try to write the tests soon.

**Comment by crookedneighbor** [comment]
No worries. I'm happy to help anyone who is willing to write tests for Habit :smiley:

**Comment by hairlessbear** [comment]
Just putting this here so I don't forget (and so anyone can correct/add as necessary):

Tests needed:
1) User is in Inn (`user.preferences.sleep = true`), cron runs, user is still in Inn.
2) User is in Inn, checks off some (but not all) dailies, cron runs, all dailies are unchecked and no damage is done
3) User is in Inn, checks off some (but not all) dailies, checks out, cron runs and unchecked dailies cause damage

**Comment by Alys** [comment]
Is this a good time to mention checklists? :) If some checklist items are ticked, cron should probably untick them.

However an exception would be for Grey Dailies where the Daily itself has not been completed. Checklist items on those Dailies should remain ticked. This is a feature that will be introduced for normal cron (when you aren't resting in the inn) when the skills rebalancing code goes live. See https://github.com/HabitRPG/habitrpg/issues/3029#issuecomment-68128382 
_"When a Grey Daily has ticked checklist items but the Daily itself is not completed, those items should NOT become unticked by cron. Checklist items should only become unticked if either the Daily is due (not grey) or if the Daily has been completed. #4736"_

However I think it would be acceptable if this PR went live in its current state (plus tests) and then checklist handling was added later (although ideally reasonably soon).

**Comment by hairlessbear** [comment]
ping @crookedneighbor I added some tests; how do they look?

@Alys those checklist tests sound like they belong in the dailies block, not the Inn block I'm writing. I'd be glad to try to add them in a different PR though :) Also, as @crookedneighbor mentioned, the whole dailies block is currently skipped for some reason. Should that be undone now/when adding checklist tests?

**Comment by Alys** [comment]
It's not just tests for checklists it's the actual unticking feature that's needed for when you're resting in the inn. But I'm happy to have it done in a separate PR! I don't think it should hold up the release of this PR.

**Comment by hairlessbear** [comment]
But the unticking feature happens regardless of whether you're in the Inn, doesn't it? If so, there will need to be tests added for the dailies and for the Inn; as far as I can tell, there currently aren't any checklist tests.

**Comment by Alys** [comment]
Yes, the unticking behaviour should be identical regardless of whether you are in the inn or not. I imagine that the same tests could be used for both cases.

**Comment by hairlessbear** [comment]
Got it. I'll try to get that in soon, but my schedule is rather insane the next couple weeks, so it might not happen for a while :/

**Comment by crookedneighbor** [comment]
@Alys the checklist behavior isn't live yet, right? It's something that's going live with the skill rebalancing? So it'd make more sense to write those tests into the skills rebalancing branch, or into develop after the skills go lives. 

By the way, sure would like some tests around the new skill changes :wink:

@hairlessbear This is a good start. There's one design change I'd like to see happen. When writing tests, it's good to think of them as documentation for the code. Your `describe` block tells you the piece of code you are testing, so in this case, "Inn", like you have it, and each `it` block tells you some expectation about it. Right now you have one `it` block that says "handles inn behavior", but that doesn't tell me much about what that behavior is. It might be more useful instead to write a couple of `it` blocks, maybe something like this:

``` coffeescript
describe "Resting in the inn", ->
    beforeEach ->
        # set up user and dailies for testing
    it "resets dailies", ->
        # expectations
    it "does not damage user for incomplete dailies", ->
        # expectations
    it "gives credit for complete dailies", ->
        # expectations
```

Does that make sense? Feel free to play around with the wording if you can think of something better.

**Comment by deilann** [comment]
I can't express how much more useful this has made Habit for me already.

Thank you so much.

**Comment by hairlessbear** [comment]
@crookedneighbor Your suggestions definitely make sense. What do you think of the tests now?

Also, I wrote the tests in the style you suggested, but the `beforeEach` didn't work at first. After some googling, I found out that, for reasons I'm still fuzzy on (something about scope), `beforeEach` needs to have variables set to `null` before the `beforeEach` block (I found that [here](http://stackoverflow.com/questions/17840747/using-jasmines-beforeeach-method)). Once I added that, it worked perfectly.

@deilann I'm glad you find this so helpful :smile: Your comment made me happy :)

**Comment by hairlessbear** [comment]
I just noticed some issues with my tests; fixing now.

**Comment by hairlessbear** [comment]
Issues fixed; the tests should be ok now (at least in theory).

**Comment by crookedneighbor** [comment]
I was going to mention some of the things you fixed, so :+1:. Going to leave just a few comments, but overall looking really good.

File: test/common/algos.mocha.coffee [hunk 1]
```diff
@@ -185,7 +185,42 @@ describe 'User', ->
...
+  describe 'Resting in the Inn', ->
+    user = null
+    cron = null
+    beforeEach ->
+      user = newUser()
+      user.preferences.sleep = true
+      cron = -> user.lastCron = moment().subtract(1, 'days');user.fns.cron()
+      user.dailys = []
+      _.times 2, -> user.dailys.push shared.taskDefaults({type:'daily'})
+    it "doesn't change on cron", ->
```
Comment (crookedneighbor) [inline test/common/algos.mocha.coffee]: I'd be more explicit about what doesn't change, maybe `it "remains in the inn on cron", ->`

File: test/common/algos.mocha.coffee [hunk 2]
```diff
@@ -185,7 +185,42 @@ describe 'User', ->
...
+      expect(user.preferences.sleep).to.be.ok
+
+    it 'resets dailies', ->
+       user.dailys[0].completed = true
+       cron()
+       expect(user.dailys[0].completed).to.not.be.ok
+
+    it 'does not damage user for incomplete dailies', ->
+      user.dailys[0].completed = true
+      user.dailys[1].completed = false
```
Comment (crookedneighbor) [inline test/common/algos.mocha.coffee:209]: I'd put an expectation here, before cron, that the user has 50 HP.

File: test/common/algos.mocha.coffee [hunk 3]
```diff
@@ -185,7 +185,42 @@ describe 'User', ->
...
 
+    it 'gives credit for complete dailies', ->
+      user.dailys[0].completed = true
+      cron()
+      expect(user.dailys[0].history.length).to.be.ok
+
+    it 'damages user for incomplete dailies after checkout', ->
+      user.dailys[0].completed = true
+      user.dailys[1].completed = false
+      user.preferences.sleep = false
```
Comment (crookedneighbor) [inline test/common/algos.mocha.coffee:223]: Same here: I'd put an expectation, before cron, that the user has 50 HP.

**Comment by crookedneighbor** [comment]
You're exactly right about scope. Remember that coffeescript just gets compiled down to javascript, so the `beforeEach, ->` just turns into:

``` javascript
beforeEach(function(){
    // Do stuff in here
});
```

So when you declare the user in the beforeEach without declaring it outside of the function first, you're limiting the scope, or what has access to those variables, to the beforeEach function. So:

``` coffeescript
beforeEach ->
      user = newUser()
```

gets compiled to:

``` javascript
beforeEach(function(){
    var user = newUser();
});
```

And only the anonymous function inside the beforeEach block has access to the variable.

If you write:

``` coffeescript
user = null
beforeEach ->
      user = newUser()
```

that gets compiled to:

``` javascript
var user;
beforeEach(function(){
    user = newUser();
});
```

And since it was declared outside of the function, the scope is the whole describe block and those it blocks have access to it. 

Does that make the fuzzy notion of scope a little clearer?

**Comment by hairlessbear** [comment]
@crookedneighbor Ah, I see where I was confused. I made some incorrect assumptions about how javascript, and, in particular, `beforeEach` worked. Thanks for the explanation; that makes a lot more sense.

I also updated the tests to your specifications; thoughts?

File: test/common/algos.mocha.coffee [hunk 4]
```diff
@@ -185,7 +185,44 @@ describe 'User', ->
...
+       expect(user.dailys[0].completed).to.not.be.ok
+
+    it 'does not damage user for incomplete dailies', ->
+      expect(user).toHaveHP 50
+      user.dailys[0].completed = true
+      user.dailys[1].completed = false
+      cron()
+      expect(user).toHaveHP 50
+
+    it 'gives credit for complete dailies', ->
```
Comment (crookedneighbor) [inline test/common/algos.mocha.coffee:213]: The rest all look awesome, I just forgot to mention this before in my last round of comments. 

Similarly to checking the hp before hand, we should also check the history length before hand. Further, I'd use the expect parameter `.empty`, so the whole thing all together would look like this:

``` coffeescript
it 'gives credit for complete dailies', ->
      user.dailys[0].completed = true
      expect(user.dailys[0].history).to.be.empty
      cron()
      expect(user.dailys[0].history).to.be.not.empty
```

Make that update and it'll be ready to merge in!

**Comment by hairlessbear** [comment]
@crookedneighbor done! I had to make one change though. `to.be.not.empty` has to be `to.not.be.empty` :)

**Comment by hairlessbear** [comment]
Or more accurately, `not`just needs to be before `be`. As far as I can tell, the order of `to` and `not` doesn't matter (at least judging by the examples [here](https://github.com/Automattic/expect.js/blob/master/README.md)).

**Comment by crookedneighbor** [comment]
:+1: I always forget the order.

**Comment by crookedneighbor** [comment]
@Alys @deilann @lemoness I'm happy with the tests and ready to merge this in. Have any of you had problems with it on your beta testing?

**Comment by lemoness** [comment]
So, @veeeeeee and I have been in the inn on beta for the past half week,
but we have yet to have it reset our Dailies. I'd assumed that maybe I was
doing something wrong every day, but it's gotten to the point where I doubt
that, especially if Vicky is having the same experience. I need to check
out of the inn today, though, because I have a lot of things to do.

**Comment by hairlessbear** [comment]
@lemoness That's so strange; I've been testing it with my test user, and it has yet to not work :confused: I have no clue why the behavior would be that inconsistent.

**Comment by deilann** [comment]
Definitely resetting for me.

Are you making sure to trigger cron on beta?

**Comment by Alys** [comment]
Any process that uses the production site with your account before you use
the beta site will trigger cron on production. This could include mobile
apps running in the background.

**Comment by crookedneighbor** [comment]
The likely culprit is the chat chrome extension.

**Comment by Alys** [comment]
_"the checklist behavior isn't live yet, right? It's something that's going live with the skill rebalancing?"_

On the production site, when you are not resting in the inn, the current cron code will untick all checklist boxes for all Dailies (regardless of whether the Dailies have been completed or not, or are scheduled or not). Ideally, the same thing should happen when you are resting in the inn, although I don't think that's important enough to hold up the release of this PR. However there'll probably be people asking for that to be added after this PR goes live; it will look a bit weird if Dailies are unticked but their checklist items are not (but that's still a much better situation than we have now on production when you are Resting).

The new behaviour that will go live with the skills rebalancing is that checklist items are NOT unticked by cron if they belong to an incomplete Grey Daily.

**Comment by hairlessbear** [comment]
@Alys I think you're right that people will wonder why the checklists are behaving differently when they're in the Inn. On top of that, since the dailies are no longer frozen, should we change the behavior of single-action habits as well? Currently, when a user is in the Inn, the single-action habits don't slowly move toward yellow as they usually do. This made sense when the dailies were frozen, but now that they won't be, it seems inconsistent. 

This behavior discrepancy is even more confusing when you consider that the `+/-` habits' functionality stays the same when a user is in the Inn.

**Comment by Alys** [comment]
If we change Habits so they aren't frozen, we'd need to do To-Dos as well (they don't get redder while you are in the Inn).

We should probably decide this: **What is Resting in the Inn for?**

Previously, it was for freezing everything so that two or more days were treated like a single day. Are we moving towards merely "Resting in the Inn is for avoiding damage from your Dailies"?

Alternatively, **Why are we making this change to the Inn?**

Are we doing this simply so that people undergoing stressful times / sickness can still use their Dailies to keep some control over their lives, or are we wanting to change the basic functions of the Inn to a larger extent?

I think I'm in favour of as few changes as possible (untick Dailies and their checklists and that's it). It's okay that Resting in the Inn gives different behaviour across the site because it's something you use only when you have a real need for it, not for regular use that makes almost all normal functions available.

**Comment by deilann** [comment]
I'd like to change Habits and To-Dos to function more like normal.

While the inn isn't meant to be for regular use, folks with chronic illness
or other reason may be using it more often than the average user. Because
of this, their Habits and To-Dos will be frozen on many days, when it
doesn't make sense for them to be.

I think the inability to get a Perfect Day buff or deal quest damage is
enough of an incentive not to be in the inn. Also taking away the system
that allows you to recognize your progress on your Habits and To-Dos is too
much.

**Comment by Alys** [comment]
@deilann Fair points. I have changed my mind about unfreezing Habits and To-Dos. :)

Are we still okay with not gaining Mana when you are Resting? I currently think you should not gain it, as another incentive to not be in the Inn when it's not necessary, but I don't have a good grasp of what it's like to have a chronic illness.

FYI, a summary of Cron's actions: http://habitrpg.wikia.com/wiki/Cron#Cron_Actions

**Comment by Alys** [comment]
Soz, you'd already mentioned Perfect Day buff. I'm editing that out of my previous post to make me look less inattentive. :)

**Comment by deilann** [comment]
I'd agree with not gaining Mana, especially if Habits and To-Dos still gain
mana, so you have some source, but not Daily-related source, as you're not
losing health from Dailies, so it makes sense you'd lose something there.

**Comment by lemoness** [comment]
That seems very logical to me! I like it :)

**Comment by Alys** [comment]
Ugh, sorry, I meant not gaining mana overnight when cron runs, while you are Resting.

I'd like all tasks to keep giving mana when you tick them off, purely for consistency and simple code, and to avoid repeated questions (I sense it would be a common question if Dailies didn't give mana but other tasks did).

**Comment by deilann** [comment]
I figured that's what you meant! Not gaining mana overnight makes perfect
sense to me.

**Comment by hairlessbear** [comment]
Just to make sure I understand this all correctly, to sum up, resting at the Inn will function identically to normal gameplay, with the following exceptions:
1) No perfect day buff (can still get buffs from party members casting spells)
2) No mana regen from cron (can still regenerate from checking off dailies and habits and party members casting _Ethereal Surge_)
3) No damage to boss or quest items collected (boss can still damage user)

Is that the consensus?

**Comment by lemoness** [comment]
I believe so!

**Comment by hairlessbear** [comment]
@lemoness Did dailies ever reset properly for you on beta? Or is that an unresolved issue?

**Comment by lemoness** [comment]
Alas, I had to leave the Inn to get things done! But I concluded that there
were so many chances for user error (notably, my app may have been running
on my phone) that it was probably not an issue.

**Comment by hairlessbear** [comment]
In that case, is there anything else that needs to be done for this to be merged?

**Comment by crookedneighbor** [comment]
@hairlessbear I wrote in the code (and added tests) to reset the checklists on dailies and then merged it in. I've notated this toward your next contributor tier. 

@lemoness We will need a mention in the next Bailey announcement about this, as it is new behavior and sure to get "bug reports".

**Comment by Alys** [comment]
@lemoness In the message about it, I think we should state very clearly up front that this doesn't change how cron calculates damage to your or your party, or how the Inn protects you and your party members. You can keep using your Dailies exactly as you have been with no chance of extra health loss. For the skills rebalancing, there were concerns about those things being different (even though they weren't), and I think this change is even more likely to prompt unnecessary worry.

**Comment by Alys** [comment]
@crookedneighbor For the code that unticks checklist items, it looks like it does that for all Dailies, regardless of whether the Daily is scheduled or not. We've already had a several posts with positive feedback about the new feature where checklist items are not unticked if the Daily was unscheduled on the previous day (a Grey Daily), and people will have started to use this to create running totals of weekly actions (search for the word "gym" in [this post](http://habitrpg.wikia.com/wiki/User_blog:LadyAlys/Skills_Rebalancing_and_Other_Changes)). If they Rest in the Inn for one night during the week, it will wipe out all their progress so far on Grey Daily checklists.

Also, are we going to change cron so that it also allows To-Dos and Habits to age? If so, do we want to do this at the same time as this cron change?

This cron change (resetting Dailies while in the Inn) is, I think, a pretty big change. Even though it's not damaging, it's still something different for the users to get used to. Do we want to spring this on them so soon after the skills rebalancing? @lemoness ?

**Comment by Alys** [comment]
In fact, the more I think about this, the more concerned I am about how the users will take it. This needs to be handled very carefully because cron and the Inn are critical components of HabitRPG and the users need to be able to rely on them, and to _feel_ that they can rely on them. Sudden large changes will damage that.

I'm aware that the change is beneficial, desirable, and has no negative aspects. I think the majority of the users will agree once they understand what the change involves.

However to understand the change, they will need to have it explained to them. They must be told that the only difference is that Dailies become unticked, and that this does NOT mean they're being damaged while they're in the Inn.

I think this should be told to them in a Bailey and social media posts a few days BEFORE this change is released. Otherwise, those who are resting in the inn will log in, see that their Dailies have become unticked, and freak out because they'll think that they and their party have taken damage. Even if they quickly realise that's not the case, it's still an unpleasant moment for them. It won't make the site look good, and we need goodwill even more than usual now while the skills change is so new. They could easily notice the unticked Dailies before they read a Bailey released at the same time if they click on the Bailey's "later" button.

It would probably also be good to adjust the message next to the Rest in the Inn button a few days before to explain that a change is coming. That will be more obvious for some users than a Bailey will be.

@crookedneighbor  can this PR and https://github.com/HabitRPG/habitrpg/commit/c81d3213665ef89be525ce949d10a34a754a187f be removed from develop until this has been discussed further?

**Comment by Alys** [comment]
@crookedneighbor can this PR andhttps://github.com/HabitRPG/habitrpg/commit/c81d3213665ef89be525ce949d10a34a754a187f be removed from develop until this has been discussed further?  (Sorry, I pinged you with the wrong name in my previous post).

**Comment by crookedneighbor** [comment]
@Alys I've pended the tests and commented out the new behavior for dailies while in the inn.

I didn't realize that the checklists not resetting on gray dailies had become live, that's why I left it out of the checklist reset code.

**Comment by Alys** [comment]
Thanks @crookedneighbor.

**Comment by lemoness** [comment]
I would really like to delay this until people have had more time to get
adjusted to skills rebalancing - at least a week. I agree with Alys that it
will be overwhelming. Part of why I wanted to get the skills rebalancing
out there sooner is that we do have a variety of changes coming up, and I
think that it is vital to space them out. Let's revisit this in a week or
two.

**Comment by crookedneighbor** [comment]
I've started work on a branch to re-implement the new behavior with the correct behavior for checklists. I'll put it in a PR when I'm done. 

@Verabird (or anyone else that can parse coffeescript) can you explain this line to me? https://github.com/HabitRPG/habitrpg/blob/develop/common/script/index.coffee#L1555-L1556

I understand that it doesn't reset checklists on dailies that are incomplete and not due for that day, but I'm struggling to understand _how_ it is doing that.

File: common/locales/en/npc.json [hunk 5]
```diff
@@ -5,8 +5,8 @@
...
     "mattBochText1": "Welcome to the Stable! I'm Matt, the beast master. Choose a pet here to venture at your side. Feed them and they'll grow into powerful steeds.",
     "daniel": "Daniel",
-    "danielText": "Welcome to the Tavern! Stay a while and meet the locals. If you need to rest (vacation? illness?), I'll set you up at the inn. While checked-in, your Dailies are frozen as-is (checked/unchecked) until the day after check-out. You will not suffer for missing them at the day's end.",
-    "danielText2": "Be warned: If you are participating in a boss quest, the boss will still damage you for your party mates' missed Dailies!",
+    "danielText": "Welcome to the Tavern! Stay a while and meet the locals. If you need to rest (vacation? illness?), I'll set you up at the Inn. While checked-in, your Dailies won't hurt you at the day's end, but you can still check them off.",
+    "danielText2": "Be warned: If you are participating in a boss quest, the boss will still damage you for your party mates' missed Dailies! Also, your own damage to the Boss (or items collected) will not be applied until you check out of the Inn.",
```
Comment (Alys) [inline common/locales/en/npc.json:9]: @crookedneighbor This text change has gone live and it is causing confusion in the Tavern. Are you able to revert that asap? I'm working at the moment so I can't.

Comment (crookedneighbor) [inline common/locales/en/npc.json:9]: Sure. Doing it now.

**Comment by hairlessbear** [comment]
@crookedneighbor I have an additional question about the code you linked. 

``` coffeescript
task.completed = false
if completed || (scheduleMisses > 0)
  _.each task.checklist, ((i)->i.completed=false;true)
```

Since `completed` is set to `false` immediately preceding the check, isn't it always false when `if completed` runs? I realize the full `if` statement may return true when `scheduleMisses` is taken into account; I'm just referring to the value of `completed` in the statement.

**Comment by Alys** [comment]
@hairlessbear `completed` and `task.completed` are not the same thing. Line 1519 assigns a value to `completed` (at that time it's value equals the value of task.completed, but it's a separate variable after that).

@crookedneighbor Follow through the code for `scheduleMisses`. For Dailies, it's assigned to 0 at line 1530, then it's incremented at 1534 if the Daily is scheduled for that day (i.e., it's not a Grey Daily). Thus at 1555, the checklist items are unticked if the user has marked the entire Daily as completed, or if the Daily was due that day. The code is complicated by the fact that multiple days can be missed, but it's simpler to understand if you ignore that and pretend that just one day is missed.

**Comment by hairlessbear** [comment]
@Alys I see; I misunderstood what javascript that code translated to. Thank you! I just discovered that you can test out coffeescript on the coffeescript website; hopefully that'll help me answer more of my own questions in the future :)

**Comment by Alys** [comment]
@lemoness In this thread, we briefly discussed allowing Habits and To-Dos to "age" while you are Resting in the Inn, just as they do when you aren't Resting (i.e., To-Dos get redder, Habits move towards yellow). See [this comment from deilann](https://github.com/HabitRPG/habitrpg/pull/5073#issuecomment-97277606). If you're in favour of it, I'll create a new issue for it.

**Comment by lemoness** [comment]
I am in favor! Go ahead.

**Pull request merged**